test/framework/clusterctl/e2e_config.go

```diff
@@ -245,6 +248,102 @@ type Files struct {
+		return fmt.Errorf("releasemarker does not support disabling the go proxy: GOPROXY=%q", os.Getenv("GOPROXY"))
+	}
+	goproxyClient := goproxy.NewClient(scheme, host)
+	for i := range c.Providers {
+		provider := &c.Providers[i]
+		for j := range provider.Versions {
+			version := &provider.Versions[j]
+			if version.Type != URLSource {
+				continue
+			}
+			// Skipping versions that are not a resolvable marker. Resolvable markers are surrounded by `{}`
+			if !strings.HasPrefix(version.Name, "{") || !strings.HasSuffix(version.Name, "}") {
+				continue
+			}
+			releaseMarker := strings.TrimLeft(strings.TrimRight(version.Name, "}"), "{")
+			ver, err := resolveReleaseMarker(releaseMarker, goproxyClient)
+			if err != nil {
+				return errors.Wrapf(err, "Failed resolving release url %q", version.Name)
+			}
+			ver = "v" + ver
+			version.Value = strings.Replace(version.Value, version.Name, ver, 1)
+			version.Name = ver
+		}
+	}
+	return nil
+}
+
+func ResolveRelease(releaseMarker string) (string, error) {
+	scheme, host, err := goproxy.GetSchemeAndHost(os.Getenv("GOPROXY"))
+	if err != nil {
+		return "", err
+	}
+	if scheme == "" || host == "" {
+		return "", fmt.Errorf("releasemarker does not support disabling the go proxy: GOPROXY=%q", os.Getenv("GOPROXY"))
+	}
+	goproxyClient := goproxy.NewClient(scheme, host)
+	return resolveReleaseMarker(releaseMarker, goproxyClient)
+}
+
+// resolveReleaseMarker resolves releaseMarker string to verion string e.g.
+// - Resolves "go://sigs.k8s.io/cluster-api@v1.0" to the latest stable patch release of v1.0.
+// - Resolves "go://sigs.k8s.io/cluster-api@latest-v1.0" to the latest patch release of v.1.0 including rc and pre releases.
+func resolveReleaseMarker(releaseMarker string, goproxyClient *goproxy.Client) (string, error) {
+	if !strings.HasPrefix(releaseMarker, "go://") {
+		return "", errors.Errorf("Unknown release marker scheme")
+	}
+
+	releaseMarker = strings.TrimPrefix(releaseMarker, "go://")
+	if releaseMarker == "" {
+		return "", errors.New("Empty release url")
+	}
+
+	gomoduleParts := strings.Split(releaseMarker, "@")
+	if len(gomoduleParts) < 2 {
+		return "", errors.Errorf("Go module or version missing")
+	}
+	gomodule := gomoduleParts[0]
+
+	includePrereleases := false
+	if strings.HasPrefix(gomoduleParts[1], "latest-") {
+		includePrereleases = true
+	}
+	version := strings.TrimPrefix(gomoduleParts[1], "latest-") + ".0"
+	version = strings.TrimPrefix(version, "v")
+	minSemVersion, err := semver.Parse(version)
+	if err != nil {
+		return "", errors.Wrapf(err, "Parsing semver for %s", version)
+	}
+	maxSemVersion := minSemVersion
+	maxSemVersion.Minor++
+
+	parsedTags, err := goproxyClient.GetVersions(context.Background(), gomodule)
+
+	if err != nil {
+		return "", err
+	}
+
+	var picked semver.Version
+	for i, tag := range parsedTags {
+		if !includePrereleases && len(tag.Pre) > 0 {
+			continue
+		}
+		if tag.LT(maxSemVersion) && tag.Minor <= minSemVersion.Minor {
```

Comment:
We are doing <= to check because we want to return stable release in following scenario:
Let’s assume user adds `go://github.com/o/r1@v1.3` and `v1.3` does not have stable release, in this case we return stable from previous minor `v1.2` .
I have added new test case that is demonstrating this scenario please check. 
And regarding major, we are checking major when we are doing this: `tag.LT(maxSemVersion)`.

Reply:
> Let’s assume user adds go://github.com/o/r1@v1.3 and v1.3 does not have stable release, in this case we return stable from previous minor v1.2 .

I think we should fail in this scenario and not silently use a different minor version. If we configure that a 1.3 version should be used it's surprising if the tests are just using 1.2 instead

Reply:
To summarize. I think we should check for equality of major and minor version.

Additionally it would be good to add a godoc comment that this picks the latest version because GetVersions returns the versions already sorted. (just to make that clearer)